runners/core-java/src/test/java/org/apache/beam/runners/core/triggers/TriggerStateMachinesTest.java

```diff
@@ -0,0 +1,198 @@
+  }
+
+  @Test
+  public void testStateMachineForAfterProcessingTime() {
+    Duration minutes = Duration.standardMinutes(94);
+    Duration hours = Duration.standardHours(13);
+
+    AfterDelayFromFirstElement trigger =
+        AfterProcessingTime.pastFirstElementInPane().plusDelayOf(minutes).alignedTo(hours);
+
+    AfterDelayFromFirstElementStateMachine machine =
+        (AfterDelayFromFirstElementStateMachine)
+            TriggerStateMachines.stateMachineForOnceTrigger(trigger);
+
+    assertThat(machine.getTimeDomain(), equalTo(TimeDomain.PROCESSING_TIME));
+
+    // This equality is function equality, but due to the structure of the code (no serialization)
+    // it is OK to check
+    assertThat(machine.getTimestampMappers(), equalTo(trigger.getTimestampMappers()));
+  }
+
+  @Test
+  public void testStateMachineForAfterWatermark() {
+    AfterWatermark.FromEndOfWindow trigger = AfterWatermark.pastEndOfWindow();
+    AfterWatermarkStateMachine.FromEndOfWindow machine =
+        (AfterWatermarkStateMachine.FromEndOfWindow)
+            TriggerStateMachines.stateMachineForOnceTrigger(trigger);
+    // No parameters, so if it doesn't crash, we win!
+  }
+
+  @Test
+  public void testDefaultTriggerTranslation() {
+    DefaultTrigger trigger = DefaultTrigger.of();
+    DefaultTriggerStateMachine machine =
+        (DefaultTriggerStateMachine) TriggerStateMachines.stateMachineForTrigger(trigger);
+    // No parameters, so if it doesn't crash, we win!
+  }
+
+  @Test
+  public void testNeverTranslation() {
+    NeverTrigger trigger = Never.ever();
+    NeverStateMachine machine =
+        (NeverStateMachine) TriggerStateMachines.stateMachineForTrigger(trigger);
+    // No parameters, so if it doesn't crash, we win!
+  }
+
+
+  //
+  // Tests for composite trigger translation
+  //
+  // These check just that translation was invoked recursively using somewhat random
+  // leaf subtriggers; by induction it all holds together. Beyond this, explicit tests
+  // of particular triggers will suffice.
+
+  private static final int ELEM_COUNT = 472;
+  private static final Duration DELAY = Duration.standardSeconds(95673);
+
+  private final OnceTrigger subtrigger1 = AfterPane.elementCountAtLeast(ELEM_COUNT);
+  private final OnceTrigger subtrigger2 =
+      AfterProcessingTime.pastFirstElementInPane().plusDelayOf(DELAY);
+
+  private final OnceTriggerStateMachine submachine1 =
+      TriggerStateMachines.stateMachineForOnceTrigger(subtrigger1);
+  private final OnceTriggerStateMachine submachine2 =
+      TriggerStateMachines.stateMachineForOnceTrigger(subtrigger2);
+
+  @Test
+  public void testAfterEachTranslation() {
+    AfterEach trigger = AfterEach.inOrder(subtrigger1, subtrigger2);
+    AfterEachStateMachine machine =
+        (AfterEachStateMachine) TriggerStateMachines.stateMachineForTrigger(trigger);
+
+    assertThat(machine, equalTo(AfterEachStateMachine.inOrder(submachine1, submachine2)));
```

Comment:
If each TriggerStateMachine has a equals method, I would consider making this a parameterized test where you can write:

```
new Object[] { AfterEach.inOrder(subtrigger1, subtrigger2), AfterEachStateMachine.inOrder(submachine1, submachine2) }
...
```